[RISCV] Avoid VMNOT by swapping VMERGE operands for mask extensions

pfusik opened the pull request:
Fold:

    (select (not m),  1, 0) -> (select m, 0,  1)
    (select (not m), -1, 0) -> (select m, 0, -1)

preames commented:
In concept, sound reasonable.  I had played with a more generic vmerge operand swap a while back, and got lost in unprofitable cases.  Something restricted specifically to the extend doesn't seem unreasonable.  Would need a real patch and tests to meaningful review beyond that.

pfusik committed:
[RISCV] Avoid VMNOT by swapping VMERGE operands for mask extensions

Fold:

    (select (not m),  1, 0) -> (select m, 0,  1)
    (select (not m), -1, 0) -> (select m, 0, -1)

preames reviewed (approved):
LGTM

Please make sure to update the review description to be non-blank before landing.

Do you have commit access?  Or do you need me to land for you?

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -8965,6 +8965,9 @@ SDValue RISCVTargetLowering::lowerVectorMaskExt(SDValue Op, SelectionDAG &DAG,
     SDValue SplatZero = DAG.getConstant(0, DL, VecVT);
     SDValue SplatTrueVal = DAG.getSignedConstant(ExtTrueVal, DL, VecVT);
+    if (Src.getOpcode() == ISD::XOR &&
+        ISD::isConstantSplatVectorAllOnes(Src.getOperand(1).getNode(), false))
```
The `BuildVectorOnly` arg defaults to false so you can omit if you'd like

pfusik replied:
Default arg omitted, thanks!

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -8980,6 +8983,20 @@ SDValue RISCVTargetLowering::lowerVectorMaskExt(SDValue Op, SelectionDAG &DAG,
+      if (ScalableOnes.getOpcode() == ISD::INSERT_SUBVECTOR &&
+          ScalableOnes.getOperand(0).isUndef() &&
+          ISD::isConstantSplatVectorAllOnes(
+              ScalableOnes.getOperand(1).getNode(), false)) {
```
Same here

pfusik replied:
Done

lukel97 reviewed (approved):
LGTM

pfusik commented:
> LGTM
> 
> Please make sure to update the review description to be non-blank before landing.

Done.
Also added negative tests for XOR that is not a NOT. Force-pushed because of editing the first commit with the tests.

> Do you have commit access? Or do you need me to land for you?

I do. Since the GitHub interface is restricted to squash merges, I'll cherry-pick the commit adding the tests to the main branch. I've done that before.

pfusik commented:
Tests merged into main branch as 97872409125bd9972ab7c74f1fe25539820f6631. PR rebased.
